### PR TITLE
[FLINK-3758] Add possibility to register accumulators in custom triggers

### PR DESCRIPTION
**Pull request by knaufk**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.
- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)
- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added
- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/functions/util/AbstractRuntimeUDFContext.java [hunk 1]
```diff
@@ -43,9 +38,13 @@
 import org.apache.flink.api.common.state.ValueStateDescriptor;
 import org.apache.flink.api.common.typeinfo.TypeInformation;
 import org.apache.flink.core.fs.Path;
-
 import static org.apache.flink.util.Preconditions.checkNotNull;
 
+import java.io.Serializable;
```
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/util/AbstractRuntimeUDFContext.java]: unrelated changes. the order of imports should stay the same, even though it might be arbitrary 😃

File: flink-core/src/main/java/org/apache/flink/api/common/functions/util/AbstractRuntimeUDFContext.java [hunk 2]
```diff
@@ -133,13 +132,24 @@ public DoubleCounter getDoubleCounter(String name) {
 		accumulators.put(name, accumulator);
 	}
 
-	@SuppressWarnings("unchecked")
```
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/util/AbstractRuntimeUDFContext.java]: This should stay to prevent warnings, should also be added to the other `getAccumulator` method.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/Trigger.java [hunk 3]
```diff
@@ -169,7 +170,17 @@ public void clear(W window, TriggerContext ctx) throws Exception {}
 		 *                                       function (function is not part os a KeyedStream).
 		 */
 		<S extends State> S getPartitionedState(StateDescriptor<S, ?> stateDescriptor);
-	
+
+		/**
+		 *
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/Trigger.java]: extra line

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/Trigger.java [hunk 4]
```diff
@@ -169,7 +170,17 @@ public void clear(W window, TriggerContext ctx) throws Exception {}
...
+		 * Get an existing accumulator by <code>name</code> if it exists, Otherwise add <code>defaultAccumulator</code>.
+		 *
+		 * @param name Name of the accumulator
+		 * @param defaultAccumulator Accumulator to add, if no accumulator of the given name exists
+		 * @return The accumulator registered by the given <code>name</code> or the given default <code>defaultAccumulator</code>
+		 */
+		public <V, A extends Serializable> Accumulator<V, A> getAccumulator(String name, org.apache.flink.api.common.accumulators.Accumulator<V, A> defaultAccumulator);
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/Trigger.java]: `org.apache.flink.api.common.accumulators.Accumulator` can be shortened to `Accumulator`

**Comment by aljoscha** [comment]
Nice and straightforward, I just had some minor comments about formatting. 😄

And of course we protect the feature with a test.

If you want I can fix the formatting while merging. Or you can fix it.

**Comment by knaufk** [comment]
Hopefully, fixed the formatting stuff. If there are still missing lines or so, please just go ahead and fix it during merging...

**Comment by aljoscha** [comment]
Ok, thanks! I'm merging it.

**Comment by aljoscha** [comment]
R: @rmetzger 
R: @StephanEwen 

What do you think about the interface. This adds a method to `RuntimeContext`. This does not break if users are only using it. It would break if someone implements it.

**Comment by knaufk** [comment]
I was worried about this to and intially only added the method to the AbstractRuntimeUDFContext. On the other hand, are you aware of anyone who has implemented RuntimeContext?

**Comment by aljoscha** [comment]
No one should implement it. With interfaces it's hard to now what the stance is, that's why I'm including those two. @rmetzger implemented the tool that will check whether releases break API compatibility and @StephanEwen always has an eye on these.

**Comment by rmetzger** [comment]
How about tagging the new method with `@PublicEvolving`. Since its a new API anyways, its good not to make it part of the stable interfaces.

With Flink 2.0, we can remove the `Evolving` part ;)

**Comment by StephanEwen** [comment]
Sorry for chiming in late.

I was wondering whether we actually need accumulators here, or what we should add is access to the metrics. Because the use case seems to be "monitoring", not "side aggregates".

Admittedly, the accumulators were the means of choice for metrics up to now, but for the future (now that we are adding proper metrics), we may want to fix this.

**Comment by aljoscha** [comment]
@knaufk Could you please close this now that the Jira issue is also closed?
